res/res_pjsip_refer.c

```diff
@@ -786,6 +791,498 @@ static void refer_blind_callback(struct ast_channel *chan, struct transfer_chann
+};
+
+/*! \brief Helper function which returns the name-addr of the Refer-To header or NULL */
+static pjsip_uri *get_refer_to_uri(pjsip_tx_data *tdata)
+{
+	const pj_str_t REFER_TO = { "Refer-To", 8 };
+	pjsip_generic_string_hdr *refer_to;
+	pjsip_uri *parsed_uri;
+
+	if (!(refer_to = pjsip_msg_find_hdr_by_name(tdata->msg, &REFER_TO, NULL))
+		|| !(parsed_uri = pjsip_parse_uri(tdata->pool, refer_to->hvalue.ptr, refer_to->hvalue.slen, 0))
+		|| (!PJSIP_URI_SCHEME_IS_SIP(parsed_uri) && !PJSIP_URI_SCHEME_IS_SIPS(parsed_uri))) {
+		return NULL;
+	}
+
+	return parsed_uri;
+}
+
+static pj_status_t refer_on_tx_request(pjsip_tx_data *tdata) {
+	RAII_VAR(struct ast_str *, refer_to_str, ast_str_create(PJSIP_MAX_URL_SIZE), ast_free_ptr);
+	const pj_str_t REFER_TO = { "Refer-To", 8 };
+	pjsip_generic_string_hdr *refer_to_hdr;
+	pjsip_dialog *dlg;
+	struct refer_data *refer_data;
+	pjsip_uri *parsed_uri;
+	pjsip_sip_uri *refer_to_uri;
+
+	/*
+	 * If this is a request in response to a 401/407 Unauthorized challenge, the
+	 * Refer-To URI has been rewritten already, so don't attempt to re-write it again.
+	 * Checking for presence of the Authorization header is not an ideal solution. We do this because
+	 * there exists some race condition where this dialog is not the same as the one used
+	 * to send the original request in which case we don't have the correct refer_data.
+	 */
+	if (!refer_to_str
+		|| pjsip_msg_find_hdr(tdata->msg, PJSIP_H_AUTHORIZATION, NULL)
+		|| !(dlg = pjsip_tdata_get_dlg(tdata))
+		|| !(refer_data = pjsip_dlg_get_mod_data(dlg, refer_out_of_dialog_module.id))
+		|| !refer_data->to_self
+		|| !(parsed_uri = get_refer_to_uri(tdata))) {
+		goto out;
+	}
+	refer_to_uri = pjsip_uri_get_uri(parsed_uri);
+	ast_sip_rewrite_uri_to_local(refer_to_uri, tdata);
+
+	pjsip_uri_print(PJSIP_URI_IN_CONTACT_HDR, parsed_uri, ast_str_buffer(refer_to_str), ast_str_size(refer_to_str));
+	refer_to_hdr = pjsip_msg_find_hdr_by_name(tdata->msg, &REFER_TO, NULL);
+	pj_strdup2(tdata->pool, &refer_to_hdr->hvalue, ast_str_buffer(refer_to_str));
+
+out:
+	return PJ_SUCCESS;
+}
+
+static int refer_unreference_dialog(void *obj)
+{
+	struct refer_out_of_dialog *data = obj;
+
+	/* This is why we keep the dialog on the subscription. When the subscription
+	 * is destroyed, there is no guarantee that the underlying dialog is ready
+	 * to be destroyed. Furthermore, there's no guarantee in the opposite direction
+	 * either. The dialog could be destroyed before our subscription is. We fix
+	 * this problem by keeping a reference to the dialog until it is time to
+	 * destroy the subscription.
+	 */
+	pjsip_dlg_dec_session(data->dlg, &refer_out_of_dialog_module);
+	data->dlg = NULL;
+
+	return 0;
+}
+/*! \brief Destructor for REFER out of dialog structure */
+static void refer_out_of_dialog_destroy(void *obj) {
+	struct refer_out_of_dialog *data = obj;
+
+	if (data->dlg) {
+		ast_sip_push_task_wait_servant(refer_serializer, refer_unreference_dialog, data);
```

Comment:
I'm fairly certain we've tried this before and it's gone extremely poorly - specifically putting a push task wait servant in a destructor. Yeah, the documentation even calls this out:

```
* \warning \b Use of this function in an ao2 destructor callback is a
 * bad idea.  You don't have control over which thread executes the
 * destructor.  Attempting to shift execution to another thread with
 * this function is likely to cause deadlock.
```

I would be extremely concerned about including it, and I'm not sure we could fully verify every thread that might execute it and the repercussions.

Reply:
That makes sense. I based this off the implementation in `res_pjsip_pubsub.c`, which does a similar thing in line 1414.

Why this was done in pjsip_pubsub is explained in 7f9823ff. I'm fairly certain that the same holds for our REFERs, since they implicitly create a subscription. So this might be a case where it is okay to do a push task wait servant in a destructor. However, you know the intricacies much better than me, so if you want me to look into other options I will keep digging into this.

Reply:
It's from 8 years ago, so to say I know the intricacies is a stretch. Hehe.

Reply:
If you've been using it this long, and there is precedence elsewhere for doing so, then it's likely a safe case. Just update with a comment stating such.

Reply:
Done.